### DISCUSSION TITLE
ImageField validation doesn't allow me save a path

### DISCUSSION THREAD

**Discussion by mgaligniana** [encode/django-rest-framework#8898] [Question & Answer]
Let's say I have this model to save an image:

```
class Profile(models.Model):
    photo = models.ImageField(max_length=255, null=True, blank=True)
```
with this serializer:
```
class ProfileSerializer(serializers.ModelSerializer):
    class Meta:
        model = Profile
        fields = ("id", "photo")
```

Then I moved to S3 using Presigned Urls: instead of upload a file, the frontend does it for me, and I just do a POST to the Profile's view with `photo` as the path which S3 gave me (i.e. uploads/photos/foo.png)

I can run `profile = Profile.objects.create(photo="uploads/photos/foo.png")` successfully and then have access to `profile.url` (MEDIA_ROOT + path) and `profile.name` (foo.png).

But when I want create it via POST I get "The submitted data was not a file. Check the encoding type on the form." because it requires send `size`: https://github.com/encode/django-rest-framework/blob/15c613a9eb645c63102b9e894199bcf1c9bf4d65/rest_framework/fields.py#L1513

- Why does`serializers.FileField` differ from the `models.FileField` validation logic?
- What would be the changes required to get the following expected result:
`POST` to `/api/v1/profiles/` with `{"photo": "foo/bar/baz.png}"` and then retrieve the profile and get `{"id": 1, "photo": "https://s3-bucket.com/foo/bar/baz.png}"`

Thank you!

**Comment by auvipy** [comment]
I think you could try https://django-storages.readthedocs.io/en/latest/backends/amazon-S3.html

**Reply by mgaligniana** [reply]
Thank you for the answer!

I didn't know about this library! No doubt if I had to do something from scratch I would use that. But in this case I was looking for a solution that doesn't involve add a library or a big change/customization because in my opinon as I'm able to save an `ImageField` just passing an url, the `serializer.ImageField`'s validation behaviour should be the same, and don't require for an extra param such as size.

I solved in this way: I defined the `photo` field as `read_only` and overrided the `ViewSet.create` method, doing:
```
def create(self, request, *args, **kwargs):
    serializer = self.get_serializer(data=request.data)
    serializer.is_valid(raise_exception=True)
    instance = serializer.save(photo=request.data.get("photo"))
    instance.save()
    return Response(
        self.serializer_class(instance).data, status=status.HTTP_201_CREATED
    )
```

What do you think?

Thank you again!

**Reply by nachollica** [reply]
hi there! i think (not sure) that you can achieve the same result in a more robust way by overriding only the `perform_create` method, like this:
```python
def perform_create(self, serializer):
    serializer.save(photo=self.request.data.get("photo"))
```

hope it helps!

**Reply by mgaligniana** [reply]
Totally agree _Nacho_ 🇦🇷 !

But as in my final code I need to manipulate the instance and the request, and `perform_create` doesn't return the instance nor receive the request as parameter, I prefer to leave it as it is.

Thank you!
